File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.py
class Solution:
    def isPossible(self, arr, n, m, minAllowed):
        currentPosition = arr[0]
        balls = 1

        for i in range(1, n):
            if arr[i] - currentPosition >= minAllowed:
                balls += 1
                currentPosition = arr[i]

            if balls == m:
                return True

        return False

    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        n = len(position)

        ans = -1
        start = 1
        end = position[-1] - position[0]

        while start <= end:
            mid = start + (end - start) // 2

            if self.isPossible(position, n, m, mid):
                ans = mid
                start = mid + 1
            else:
                end = mid - 1

        return ans

File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.java
class Solution {
    public boolean isPossible(int [] arr, int n, int m, int minAllowed){
        int currentPosition = arr[0];
        int balls = 1;
        for(int i = 1; i < n; i++){
            if(arr [i] - currentPosition >= minAllowed){
                balls++;
                currentPosition = arr [i];
            }
            if(balls == m){
                return true;
            }

        }
        return false;
    }
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int n = position.length;
        int ans = -1;
        int start = 1;
        int end = position [n - 1] - position [0];
        while(start <= end){
            int mid = start + (end - start) / 2;
            if(isPossible(position, n, m, mid)){
                ans = mid;
                start = mid + 1;
            }
            else{
                end = mid  - 1;
            }
        }
        return ans;
    }
}